Move other div and full variants to upgraders map

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77574
* __->__ #73586

tugsbayasgalan committed:
Move other div variants to upgraders map

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34557045).

tugsbayasgalan committed:
Update on "Move other div variants to upgraders map"


Differential Revision: [D34557045](https://our.internmc.facebook.com/intern/diff/D34557045)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34557045).

gmagogsfm commented on torch/csrc/jit/operator_upgraders/version_map.cpp:
```diff
@@ -36,22 +36,42 @@ static std::unordered_map<std::string, std::vector<UpgraderEntry>> operatorVersi
       {{4,
         "div_Tensor_0_3",
         "aten::div.Tensor(Tensor self, Tensor other) -> Tensor"}}},
+     {"aten::div.Tensor_mode",
```
How about `full` operator?

tugsbayasgalan replied:
good point! included

tugsbayasgalan committed:
Update on "Move other div variants to upgraders map"


Differential Revision: [D34557045](https://our.internmc.facebook.com/intern/diff/D34557045)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34557045).

tugsbayasgalan committed:
Update on "Move other div and full variants to upgraders map"


Differential Revision: [D34557045](https://our.internmc.facebook.com/intern/diff/D34557045)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34557045).

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -751,15 +751,28 @@ struct PythonPrintImpl {
+        // TODO (tugsuu) this needs to be removed once we think it is safe.
+        uint64_t historic_version = get_min_version_for_kind(node->kind());
+
+        if (historic_version != 0) {
```
Could you also add a flag to revert to old behavior? In this error message, let's also let user know that they can use the flag to make it work.

gmagogsfm replied:
Maybe not a flag, since it is harder to use in a notebook. Maybe some new private API called `torch.use_old_upgrader_version`. Up to you.

tugsbayasgalan replied:
Are you saying we should deprecate the ENABLE_UPGRADERS macro flag and let user flip it through API? Or are you thinking about adding another flag? If latter, I am afraid this logic is getting overly complex with different flags.

gmagogsfm replied:
Not deprecating the `ENABLE_UPGRADERS` macro. But a simple way for users to revert to old `min_version_ = std::max(min_version_, get_min_version_for_kind(node->kind()));` behavior rather than relying solely on upgraders to compute version to write. `ENABLE_UPGRADERS` is difficult to toggle by an end user.

tugsbayasgalan committed:
Update on "Move other div and full variants to upgraders map"


Differential Revision: [D34557045](https://our.internmc.facebook.com/intern/diff/D34557045)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/operator_upgraders/version_map.cpp:
```diff
@@ -96,5 +120,15 @@ void test_only_reset_flag() {
   isVersionMapSorted = false;
 }
 
+static bool calculatePackageVersionBasedOnUpgraders = false;
```
Could you remind me what this is for?

tugsbayasgalan replied:
The temp fix we did was to still calculate version number for the package from the legacy map Mike introduced so that the overloads for div and full, that were not in our upgrader map, would be upgraded to correct version number. Since we figured it was risky to roll out a change involving version number, this flag is used to safely switch between the temp hack and the more correct approach where we add all the overloads in the version map. The end user can toggle this flag in the case of emergency.

gmagogsfm replied:
Ah, I remember now, just got thrown off by the name of "package"

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -753,15 +753,21 @@ struct PythonPrintImpl {
-        min_version_ = std::max(
-            min_version_, uint64_t(entry[entry.size() - 1].bumped_at_version));
+        uint64_t current_version = entry[entry.size() - 1].bumped_at_version;
+        uint64_t historic_version = get_min_version_for_kind(node->kind());
```
```suggestion
        uint64_t legacy_version_map_version = get_min_version_for_kind(node->kind());
```

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -753,15 +753,21 @@ struct PythonPrintImpl {
+        // True means we solely calculate based on upgrader version
+        if (get_version_calculator_flag()) {
+          min_version_ = std::max(min_version_, current_version);
+        } else {
```
I think we briefly discussed about this. Should we error out here if `min_version` and `historic_version` do not match? Since this error happens at compile/save time, it is a good place to detect issues. Also provide a way to override this check and use legacy versioning method

tugsbayasgalan replied:
Hmm it seems weird to error out based on that because this is an implementation detail. And this check would only work for 'div' and 'full' because historic (legacy) map doesn't have the other ops.

gmagogsfm replied:
Good point, it is up to you. I was just thinking about it as a temporary measure to make sure the migration is absolutely safe.

tugsbayasgalan committed:
Update on "Move other div and full variants to upgraders map"


Differential Revision: [D34557045](https://our.internmc.facebook.com/intern/diff/D34557045)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Move other div and full variants to upgraders map"


Differential Revision: [D34557045](https://our.internmc.facebook.com/intern/diff/D34557045)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Move other div and full variants to upgraders map"


Differential Revision: [D34557045](https://our.internmc.facebook.com/intern/diff/D34557045)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge this please
